tests/test_external_strategies.py

```diff
@@ -0,0 +1,119 @@
+"""
+Copyright 2018 Grid Singularity
+This file is part of D3A.
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+from unittest.mock import MagicMock, Mock
+
+from d3a.models.area import Area
+from d3a.models.strategy.external_strategies.load import LoadForecastExternalStrategy
+from d3a.models.strategy.external_strategies.pv import PVForecastExternalStrategy
+from d3a_interface.constants_limits import GlobalConfig
+from pendulum import now, Duration
+from pytest import mark
+
+
+def ext_strategy_fixture(strategy):
+    config = MagicMock()
+    config.ticks_per_slot = 90
+    GlobalConfig.end_date = GlobalConfig.start_date + Duration(days=1)
+    area = Area(name="forecast_pv", config=config, strategy=strategy,
+                external_connection_available=True)
+    parent = Area(name="parent_area", children=[area])
+    parent.activate()
+    area.activate()
+    strategy.connected = True
+    market = MagicMock()
+    market.time_slot = GlobalConfig.start_date
+    parent.get_future_market_from_id = lambda _: market
+    area.get_future_market_from_id = lambda _: market
+    return strategy
+
+
+class TestPVForecastExternalStrategy:
+
+    @mark.parametrize("strategy", [ext_strategy_fixture(LoadForecastExternalStrategy()),
+                                   ext_strategy_fixture(PVForecastExternalStrategy())])
+    def test_clear_energy_buffers(self, strategy):
+        strategy.energy_forecast_buffer = {now(): 1}
+        strategy.energy_measurement_buffer = {now(): 1}
+        strategy._clear_energy_buffers()
+        assert strategy.energy_forecast_buffer == {}
+        assert strategy.energy_measurement_buffer == {}
+
+    @mark.parametrize("strategy", [ext_strategy_fixture(LoadForecastExternalStrategy()),
+                                   ext_strategy_fixture(PVForecastExternalStrategy())])
+    def test_update_energy_forecast_measurement_is_called_on_market_cycle(self, strategy):
+        strategy.update_energy_forecast = Mock()
+        strategy.update_energy_measurement = Mock()
+        strategy._clear_energy_buffers = Mock()
+        strategy.event_market_cycle()
+        strategy.update_energy_forecast.assert_called_once()
+        strategy.update_energy_measurement.assert_called_once()
+        strategy._clear_energy_buffers.assert_called_once()
+
+    def test_update_forecast_calls_set_available_energy(self, strategy=ext_strategy_fixture(
```

Comment:
```suggestion
    def test_update_energy_forecast_calls_set_available_energy(self, strategy=ext_strategy_fixture(
```

Reply:
Nit: since you're also testing `set_available_energy.assert_not_called()`, the method name is saying something a bit different than its content. Minor anyway.

Reply:
I had this name before but I spend like 15 mins in order to comply to pylint. I could not find a way to put this function header into multiple lines.

Reply:
I think this should work, right?
```python
    def test_update_energy_forecast_calls_set_available_energy(
            self, strategy=ext_strategy_fixture(PVForecastExternalStrategy())):
        """Some docstring."""
        time = strategy.area.next_market.time_slot
        # ...
```

Reply:
yeah, I realised also by myself (see my latest commits).